src/main/java/uk/gov/companieshouse/overseasentitiesapi/interceptor/UserAuthenticationInterceptor.java

```diff
@@ -74,4 +101,81 @@ public boolean preHandle(@NonNull HttpServletRequest request, @NonNull HttpServl
     protected Optional<TokenPermissions> getTokenPermissions(HttpServletRequest request) {
         return AuthorisationUtil.getTokenPermissions(request);
     }
+
+    private boolean doRoeUpdateJourneyChecks(HttpServletRequest request, HttpServletResponse response, String reqId, String companyNumberInScope,
+                                             String transactionId, Map<String, Object> authInfoMap) {
+        boolean isValidOeNumber = companyNumberInScope.startsWith("OE");
+        if (isValidOeNumber) {
+            boolean doCompanyNumbersMatch = doCompanyNumbersMatch(request, response, companyNumberInScope, transactionId);
+            authInfoMap.put("roe_update_journey_company_numbers_match", doCompanyNumbersMatch);
+            if (doCompanyNumbersMatch) {
+                ApiLogger.debugContext(reqId, "UserAuthenticationInterceptor authorised with company_oe_annual_update=create permission",
+                        authInfoMap);
+                return true;
+            } else {
+                ApiLogger.errorContext(reqId, "UserAuthenticationInterceptor unauthorised for ROE Update Journey as Company Numbers do not match", null, authInfoMap);
+                if (HttpServletResponse.SC_INTERNAL_SERVER_ERROR != response.getStatus()) {
+                    response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+                }
+                return false;
+            }
+        } else {
+            ApiLogger.debugContext(reqId, "UserAuthenticationInterceptor unauthorised for ROE Update Journey as Company Number is not a valid OE",
+                    authInfoMap);
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+    }
+
+    private boolean isRequestRoeUpdate(String companyNumberInScope) {
+        // Further check for OE number is done in doRoeUpdateJourneyChecks method
+        return StringUtils.isNotEmpty(companyNumberInScope);
+    }
+
+    private String getCompanyNumberInScope (HttpServletRequest request) {
+        final Map<String, List<String>> privileges = getERICTokenPermissions(request);
+        if (privileges.containsKey(COMPANY_NUMBER_KEY)) {
+            return privileges.get(COMPANY_NUMBER_KEY).get(0);
+        }
+        return null;
+    }
+
+    private boolean doCompanyNumbersMatch(HttpServletRequest request, HttpServletResponse response, String companyNumberInScope, String transactionId) {
+        String companyNumberInTransaction = getCompanyNumberInTransaction(request, response, transactionId);
+        if (StringUtils.isNotBlank(companyNumberInTransaction)) {
+            return companyNumberInTransaction.equalsIgnoreCase(companyNumberInScope);
+        }
```

Comment:
set response code? And rename method ensureCompanyNumbersMatch()

Reply:
The name of this method - `doCompanyNumbersMatch` - sounds ok to me based on the current implementation as it checks if the numbers match and returns a true/false value. Or are you suggesting that the implementation changes too @AndreKramer7654 ?

Reply:
yes, just to set the response status here. So check on caller can be avoided. Not renaming method is fine.